File: manual/data/python/src/function_Whoosh_terms.py
import os
from whoosh import index
from whoosh.fields import Schema, TEXT

class WhooshTerms:
    def manage_index_and_list_terms(self, index_dir: str, field_name: str = "content", add_documents: bool = True) -> str:
        if not os.path.exists(index_dir):
            if add_documents:
                os.mkdir(index_dir)
            else:
                raise ValueError(f"Index directory '{index_dir}' does not exist")
        
        if not index.exists_in(index_dir):
            if not add_documents:
                raise ValueError(f"No index found in directory '{index_dir}'")
            schema = Schema(content=TEXT(stored=True))
            ix = index.create_in(index_dir, schema)
            if add_documents:
                writer = ix.writer()
                writer.add_document(content="Hello world! This is a test document.")
                writer.add_document(content="Another test document is here.")
                writer.add_document(content="Whoosh makes text search easy.")
                writer.commit()
        else:
            ix = index.open_dir(index_dir)

        terms_list = []
        with ix.searcher() as searcher:
            field_terms = searcher.reader().iter_field(field_name)
            for term_bytes, term_info in field_terms:
                term = term_bytes.decode("utf-8")
                terms_list.append(term)

        ix.close()
        
        return ", ".join(terms_list)

File: code/generate_strategies/IR(pseudocode)/java/java/function_Whoosh_terms.java
package com.example;

import org.apache.lucene.analysis.standard.StandardAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.TextField;
import org.apache.lucene.index.*;
import org.apache.lucene.queryparser.classic.ParseException;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.FSDirectory;

import java.io.IOException;
import java.nio.file.Paths;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Collections;

public class WhooshTerms {

    public String manage_index_and_list_terms(String indexDir, String fieldName, boolean addDocuments) {
        try {
            File indexDirectory = new File(indexDir);

            if (!indexDirectory.exists()) {
                if (addDocuments) {
                    indexDirectory.mkdir();
                } else {
                    throw new IllegalArgumentException("Index directory '" + indexDir + "' does not exist");
                }
            }

            Directory directory = FSDirectory.open(Paths.get(indexDir));
            IndexReader reader = null;
            try {
                reader = DirectoryReader.open(directory);
            } catch (IOException e) {
                
                reader = null;
            }

            if (reader == null) {
                if (!addDocuments) {
                    throw new IllegalArgumentException("No index found in directory '" + indexDir + "'");
                }

                StandardAnalyzer analyzer = new StandardAnalyzer();
                IndexWriterConfig config = new IndexWriterConfig(analyzer);
                IndexWriter iwriter = new IndexWriter(directory, config);

                if (addDocuments) {
                    Document doc1 = new Document();
                    doc1.add(new TextField("content", "Hello world! This is a test document.", Field.Store.YES));
                    iwriter.addDocument(doc1);

                    Document doc2 = new Document();
                    doc2.add(new TextField("content", "Another test document is here.", Field.Store.YES));
                    iwriter.addDocument(doc2);

                    Document doc3 = new Document();
                    doc3.add(new TextField("content", "Whoosh makes text search easy.", Field.Store.YES));
                    iwriter.addDocument(doc3);

                    iwriter.commit();
                }
                iwriter.close();
                reader = DirectoryReader.open(directory);

            }
            IndexSearcher searcher = new IndexSearcher(reader);

            List<String> termsList = new ArrayList<>();
            for (LeafReaderContext leafContext : reader.leaves()) {
                LeafReader leafReader = leafContext.reader();

                Terms terms = leafReader.terms(fieldName);
                if (terms != null) {
                    TermsEnum termsEnum = terms.iterator();
                    while (termsEnum.next() != null) {
                        termsList.add(termsEnum.term().utf8ToString());
                    }
                }
            }


            reader.close();
            directory.close();

            return String.join(", ", termsList);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}